### PR TITLE
Distributed Lock API updates

### PR DESCRIPTION
**Pull request by msfussell**
## Description
The structural changes in the docs to introduce the distributed lock API and components. The howto still needs completing

## Issue reference
https://github.com/dapr/docs/issues/2485

### CODE REVIEW

**Comment by msfussell** [comment]
@yaron2 - Can you review the above. Then issue a new PR to update the HowTo for locks and is there an HTTP API for locks in this release?

File: daprdocs/content/en/developing-applications/building-blocks/distributed-lock/distributed-lock-api-overview.md [hunk 1]
```diff
@@ -0,0 +1,31 @@
+---
+type: docs
+title: "Distributed lock overview"
+linkTitle: "Overview"
+weight: 1000
+description: "Overview of the distributed lock API building block"
+---
+
+## Introduction
+Locks are used to provide mutually exclusive access to a resource. For example, you can use a lock to provide exclusive access to a database row, a table or even an entire database. Or a lock could be used to lock reading message from a queue in a sequential manner. Any resource that is shared where updates occur can be the target for a lock. Locks are not normally used on reads, only on operations that mutate state.
```
Comment (hhunter-ms) [inline daprdocs/content/en/developing-applications/building-blocks/distributed-lock/distributed-lock-api-overview.md]:
```suggestion
Locks are used to provide mutually exclusive access to a resource. For example, you can use a lock to:

- Provide exclusive access to a database row, table, or an entire database
- Lock reading messages from a queue in a sequential manner

Any resource that is shared where updates occur can be the target for a lock. Locks are usually used on operations that mutate state, not on reads.
```

> Any resource that is shared where updates occur can be the target for a lock.

Not sure I understand this sentence enough to reword it?

File: daprdocs/content/en/developing-applications/building-blocks/distributed-lock/distributed-lock-api-overview.md [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+
+A lock has a name and it is up to the application to determine the resources that this named lock accesses. Typically you have multiple instance of the same application (multiple applications with the same app ID) which use this named lock to exclusively access the resource to perform updates. For example, in the competing consumer pattern, multiple instances of an application access a queue amd you could decide that you want to lock the queue whilst the application is running its business logic.
```
Comment (hhunter-ms) [inline daprdocs/content/en/developing-applications/building-blocks/distributed-lock/distributed-lock-api-overview.md]:
```suggestion
Each lock has a name. The application determines the resources that the named lock accesses. Typically, multiple instances of the same application (multiple applications with the same app ID) use this named lock to exclusively access the resource and perform updates. 

For example, in the competing consumer pattern, multiple instances of an application access a queue. You can decide that you want to lock the queue while the application is running its business logic.
```

File: daprdocs/content/en/developing-applications/building-blocks/distributed-lock/distributed-lock-api-overview.md [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+
+The diagram belows shows an example of two instances of the same application, `App1`, using the [Redis lock component]({{< ref redis-lock >}}) to take a lock on a shared resource. The first app intance acquires the named lock, and gets exclusive access to the resource for updates. The second app instance is unable to aquire the lock and therefore is not allowed to access the resource, until the lock is released, either explicitly by the application through the unlock API or is released after a period of time, due to a lease timeout.
```
Comment (hhunter-ms) [inline daprdocs/content/en/developing-applications/building-blocks/distributed-lock/distributed-lock-api-overview.md]:
```suggestion
In the diagram below, two instances of the same application, `App1`, use the [Redis lock component]({{< ref redis-lock >}}) to take a lock on a shared resource. 

- The first app instance acquires the named lock and gets exclusive access to the resource for updates. 
- The second app instance is unable to acquire the lock and therefore is not allowed to access the resource until the lock is released, either:
   - Explicitly by the application through the unlock API, or 
   - After a period of time, due to a lease timeout. 
```

File: daprdocs/content/en/developing-applications/building-blocks/distributed-lock/distributed-lock-api-overview.md [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+
+*This API is currently in `Alpha` state.
+
+## Features
+
+### Mutually exclusive access to a resource
+At any given moment, only one application can hold a named lock.
+
+### Deadlock free using leases
+Dapr distributed locks use a lease-based locking mechanism. If an application acquires a lock and then encounters an exception and cannot free the lock, the lock is automatically released after a period of time using a lease. This prevents resource deadlocks in the event of application failures.
```
Comment (hhunter-ms) [inline daprdocs/content/en/developing-applications/building-blocks/distributed-lock/distributed-lock-api-overview.md]:
```suggestion
Dapr distributed locks use a lease-based locking mechanism. If an application acquires a lock, encounters an exception, and cannot free the lock, the lock is automatically released after a period of time using a lease. This prevents resource deadlocks in the event of application failures.
```

File: daprdocs/content/en/reference/components-reference/supported-locks/redis-lock.md [hunk 5]
```diff
@@ -0,0 +1,109 @@
+---
+type: docs
+title: "Redis"
+linkTitle: "Redis"
+description: Detailed information on the Redis lock component
+---
+
+## Component format
+
+To setup Redis lock create a component of type `lock.redis`. See [this guide]({{< ref "howto-use-distributed-lock" >}}) on how to create a lock.
```
Comment (hhunter-ms) [inline daprdocs/content/en/reference/components-reference/supported-locks/redis-lock.md]:
```suggestion
To set up the Redis lock, create a component of type `lock.redis`. See [this guide]({{< ref "howto-use-distributed-lock" >}}) on how to create a lock.
```

File: daprdocs/content/en/reference/components-reference/supported-locks/redis-lock.md [hunk 6]
```diff
@@ -0,0 +1,109 @@
...
+  - name: redisHost
+    value: <HOST>
+  - name: redisPassword
+    value: <PASSWORD>
+
+
+```
+
+{{% alert title="Warning" color="warning" %}}
+The above example uses secrets as plain strings. It is recommended to use a secret store for the secrets as described [here]({{< ref component-secrets.md >}}).
```
Comment (hhunter-ms) [inline daprdocs/content/en/reference/components-reference/supported-locks/redis-lock.md]:
```suggestion
The above example uses secrets as plain strings. It is recommended to use a secret store for the secrets, as described [here]({{< ref component-secrets.md >}}).
```

File: daprdocs/content/en/reference/components-reference/supported-locks/redis-lock.md [hunk 7]
```diff
@@ -0,0 +1,109 @@
...
+
+| Field              | Required | Details | Example |
+|--------------------|:--------:|---------|---------|
+| redisHost          | Y        | Connection-string for the redis host  | `localhost:6379`, `redis-master.default.svc.cluster.local:6379`
+| redisPassword      | Y        | Password for Redis host. No Default. Can be `secretKeyRef` to use a secret reference  | `""`, `"KeFg23!"`
+
+
+## Setup Redis
+
+Dapr can use any Redis instance - containerized, running on your local dev machine, or a managed cloud service.
```
Comment (hhunter-ms) [inline daprdocs/content/en/reference/components-reference/supported-locks/redis-lock.md]:
```suggestion
Dapr can use any Redis instance: containerized, running on your local dev machine, or a managed cloud service.
```

File: daprdocs/content/en/reference/components-reference/supported-locks/redis-lock.md [hunk 8]
```diff
@@ -0,0 +1,109 @@
...
+        metadata:
+        - name: redisPassword
+          value: lhDOkwTlp0
+    ```
+{{% /codetab %}}
+
+{{% codetab %}}
+**Note**: this approach requires having an Azure Subscription.
+
+1. Open [this link](https://ms.portal.azure.com/#create/Microsoft.Cache) to start the Azure Cache for Redis  creation flow. Log in if necessary.
```
Comment (hhunter-ms) [inline daprdocs/content/en/reference/components-reference/supported-locks/redis-lock.md]:
```suggestion
1. [Start the Azure Cache for Redis creation flow](https://ms.portal.azure.com/#create/Microsoft.Cache). Log in if necessary.
```

File: daprdocs/content/en/reference/components-reference/supported-locks/redis-lock.md [hunk 9]
```diff
@@ -0,0 +1,109 @@
...
+2. Fill out necessary information and **check the "Unblock port 6379" box**, which will allow us to persist state without SSL.
+3. Click "Create" to kickoff deployment of your Redis instance.
+4. Once your instance is created, you'll need to grab the Host name (FQDN) and your access key.
```
Comment (hhunter-ms) [inline daprdocs/content/en/reference/components-reference/supported-locks/redis-lock.md]:
```suggestion
4. Once your instance is created, you'll need to grab the Host name (FQDN) and your access key:
```

File: daprdocs/content/en/reference/components-reference/supported-locks/redis-lock.md [hunk 10]
```diff
@@ -0,0 +1,109 @@
...
+   - for the Host name navigate to the resources "Overview" and copy "Host name"
```
Comment (hhunter-ms) [inline daprdocs/content/en/reference/components-reference/supported-locks/redis-lock.md]:
```suggestion
   - For the Host name: navigate to the resource's "Overview" and copy "Host name".
```

File: daprdocs/content/en/reference/components-reference/supported-locks/redis-lock.md [hunk 11]
```diff
@@ -0,0 +1,109 @@
...
+   - for your access key navigate to "Access Keys" under "Settings" and copy your key.
```
Comment (hhunter-ms) [inline daprdocs/content/en/reference/components-reference/supported-locks/redis-lock.md]:
```suggestion
   - For your access key: navigate to "Settings" > "Access Keys" to copy and save your key.
```

File: daprdocs/content/en/reference/components-reference/supported-locks/redis-lock.md [hunk 12]
```diff
@@ -0,0 +1,109 @@
...
+5. Finally, we need to add our key and our host to a `redis.yaml` file that Dapr can apply to our cluster. If you're running a sample, you'll add the host and key to the provided `redis.yaml`. If you're creating a project from the ground up, you'll create a `redis.yaml` file as specified in [Configuration](#configuration). Set the `redisHost` key to `[HOST NAME FROM PREVIOUS STEP]:6379` and the `redisPassword` key to the key you copied in step 4. **Note:** In a production-grade application, follow [secret management]({{< ref component-secrets.md >}}) instructions to securely manage your secrets.
```
Comment (hhunter-ms) [inline daprdocs/content/en/reference/components-reference/supported-locks/redis-lock.md]:
```suggestion
5. Add your key and your host to a `redis.yaml` file that Dapr can apply to your cluster. 
   - If you're running a sample, add the host and key to the provided `redis.yaml`. 
   - If you're creating a project from the ground up, create a `redis.yaml` file as specified in [Configuration](#configuration). 
   
   Set the `redisHost` key to `[HOST NAME FROM PREVIOUS STEP]:6379` and the `redisPassword` key to the key you saved earlier. 
   
   **Note:** In a production-grade application, follow [secret management]({{< ref component-secrets.md >}}) instructions to securely manage your secrets.
```

**Review by hhunter-ms — DISMISSED** [review]

File: daprdocs/content/en/developing-applications/building-blocks/distributed-lock/distributed-lock-api-overview.md [hunk 13]
```diff
@@ -0,0 +1,43 @@
...
+
+## Introduction
+Locks are used to provide mutually exclusive access to a resource. For example, you can use a lock to:
+
+- Provide exclusive access to a database row, table, or an entire database
+- Lock reading messages from a queue in a sequential manner
+
+Any resource that is shared where updates occur can be the target for a lock. Locks are usually used on operations that mutate state, not on reads.
+
+Each lock has a name. The application determines the resources that the named lock accesses. Typically, multiple instances of the same application (multiple applications with the same app ID) use this named lock to exclusively access the resource and perform updates.
```
Comment (yaron2) [inline daprdocs/content/en/developing-applications/building-blocks/distributed-lock/distributed-lock-api-overview.md]: Remove `(multiple applications with the same app ID)`

Comment (yaron2) [reply daprdocs/content/en/developing-applications/building-blocks/distributed-lock/distributed-lock-api-overview.md]: Remove `(multiple applications with the same app ID)`

File: daprdocs/content/en/developing-applications/building-blocks/distributed-lock/distributed-lock-api-overview.md [hunk 14]
```diff
@@ -0,0 +1,43 @@
...
+
+Any resource that is shared where updates occur can be the target for a lock. Locks are usually used on operations that mutate state, not on reads.
+
+Each lock has a name. The application determines the resources that the named lock accesses. Typically, multiple instances of the same application (multiple applications with the same app ID) use this named lock to exclusively access the resource and perform updates. 
+
+For example, in the competing consumer pattern, multiple instances of an application access a queue. You can decide that you want to lock the queue while the application is running its business logic.
+
+In the diagram below, two instances of the same application, `App1`, use the [Redis lock component]({{< ref redis-lock >}}) to take a lock on a shared resource. 
+
+- The first app instance acquires the named lock and gets exclusive access to the resource for updates.
```
Comment (yaron2) [inline daprdocs/content/en/developing-applications/building-blocks/distributed-lock/distributed-lock-api-overview.md]: Remove `to the resource for updates.`

File: daprdocs/content/en/developing-applications/building-blocks/distributed-lock/distributed-lock-api-overview.md [hunk 15]
```diff
@@ -0,0 +1,43 @@
...
+   - After a period of time, due to a lease timeout. 
+
+<img src="/images/lock-overview.png" width=900>
+
+*This API is currently in `Alpha` state.
+
+## Features
+
+### Mutually exclusive access to a resource
+At any given moment, only one application can hold a named lock.
```
Comment (yaron2) [inline daprdocs/content/en/developing-applications/building-blocks/distributed-lock/distributed-lock-api-overview.md]: Replace with: `At any given moment, only one instance of an application can hold a named lock. Locks are scoped to a Dapr app-id`.

File: daprdocs/content/en/developing-applications/building-blocks/distributed-lock/howto-use-distributed-locks.md [hunk 16]
```diff
@@ -0,0 +1,64 @@
+---
```
Comment (yaron2) [inline daprdocs/content/en/developing-applications/building-blocks/distributed-lock/howto-use-distributed-locks.md]: Remove this entire file, it's a duplicate of the one above.

File: daprdocs/content/en/reference/components-reference/supported-locks/redis-lock.md [hunk 17]
```diff
@@ -0,0 +1,115 @@
...
+
+
+```
+
+{{% alert title="Warning" color="warning" %}}
+The above example uses secrets as plain strings. It is recommended to use a secret store for the secrets, as described [here]({{< ref component-secrets.md >}}).
+{{% /alert %}}
+
+
+## Spec metadata fields
```
Comment (yaron2) [inline daprdocs/content/en/reference/components-reference/supported-locks/redis-lock.md:79]: All the metadata fields of `state.redis` apply here. Same for the YAML example above for `Component format`.

**Review by yaron2 — CHANGES_REQUESTED** [review]

File: daprdocs/content/en/developing-applications/building-blocks/distributed-lock/distributed-lock-api-overview.md (hunk 13 above)
Comment (msfussell) [reply daprdocs/content/en/developing-applications/building-blocks/distributed-lock/distributed-lock-api-overview.md]: Fixed

File: daprdocs/content/en/developing-applications/building-blocks/distributed-lock/distributed-lock-api-overview.md (hunk 14 above)
Comment (msfussell) [reply daprdocs/content/en/developing-applications/building-blocks/distributed-lock/distributed-lock-api-overview.md]: Fixed

File: daprdocs/content/en/developing-applications/building-blocks/distributed-lock/distributed-lock-api-overview.md (hunk 15 above)
Comment (msfussell) [reply daprdocs/content/en/developing-applications/building-blocks/distributed-lock/distributed-lock-api-overview.md]: Fixed

File: daprdocs/content/en/developing-applications/building-blocks/distributed-lock/howto-use-distributed-locks.md (hunk 16 above)
Comment (msfussell) [reply daprdocs/content/en/developing-applications/building-blocks/distributed-lock/howto-use-distributed-locks.md]: removed

File: daprdocs/content/en/reference/components-reference/supported-locks/redis-lock.md (hunk 17 above)
Comment (msfussell) [reply daprdocs/content/en/reference/components-reference/supported-locks/redis-lock.md:79]: Pulled in all relevant fields

**Review by yaron2 — APPROVED** [review]

**Pull request merged**